pym/bob/cmds/graph.py

```diff
@@ -0,0 +1,379 @@
+import argparse
+import json
+import re
+import collections
+from ..input import RecipeSet
+
+def collect(package, excludes, done, nodes, links, maxdepth, type, level = 0):
+    def getPackageLevel(package, level):
+        if package.getRecipe().isRoot():
+            return 1
+        elif level == 0:
+            return 1
+        return 2
+
+    def getHover(package):
+        hover = collections.OrderedDict(package.getMetaEnv())
+        if package.getCheckoutStep().isValid():
+            scmidx = 0
+            for s in package.getCheckoutStep().getScmList():
+                p = s.getProperties()
+                hover['scm'+str(scmidx)] = p[0].get('scm')
+                hover['url'+str(scmidx)] = p[0].get('url')
+                scmidx = scmidx + 1
+
+        return hover
+
+    for e in excludes:
+        if (e.match(package.getName())):
+            return
+
+    if package.getPackageStep().isValid():
+        if package._getId() not in done:
+            done.add(package._getId())
+            nodes.append({"id" : package._getId(),
+                "label" : package.getName(),
+                "level" : getPackageLevel(package,level),
+                "hover" : getHover(package)})
+
+    if (maxdepth != -1) and (level >= maxdepth):
+        return
+
+    for d in package.getDirectDepSteps():
+        if (d.getPackage()._getId(), package._getId()) not in links:
+            for e in excludes:
+                if (e.match(d.getPackage().getName())):
+                    break
+            else:
+                if (type == 'dot'):
```

Comment:
That's not so elegant. Maybe this whole function could be rewritten as a generator function by `yield`'ing the packages?